staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go

```diff
@@ -291,32 +288,35 @@ func (f *fakeTimer) C() <-chan time.Time {
 func (f *fakeTimer) Stop() bool {
 	f.fakeClock.lock.Lock()
 	defer f.fakeClock.lock.Unlock()
-
+	stopped := false
 	newWaiters := make([]fakeClockWaiter, 0, len(f.fakeClock.waiters))
-	for i := range f.fakeClock.waiters {
-		w := &f.fakeClock.waiters[i]
-		if w != &f.waiter {
-			newWaiters = append(newWaiters, *w)
+	for _, w := range f.fakeClock.waiters {
+		if w.destChan == f.waiter.destChan {
+			stopped = true
+		} else {
+			newWaiters = append(newWaiters, w)
 		}
 	}
 
 	f.fakeClock.waiters = newWaiters
 
-	return !f.waiter.fired
+	return stopped
```

Comment:
Stop() is supposed to return false if it was already stopped, or if the timer had already fired. I'm having a hard time understanding how the channel equality comparison is accomplishing the latter.

Reply:
When the timer fires or is stopped, it is removed from the clock's list.  The channel equality test is how the timer's entry is identified in the list.  The fakeClockWaiter does not retain a constant memory address, so a pointer to a fakeClockWaiter can not be used to identify the one belonging to the timer.

Reply:
See if the new revision includes comments that answer this question.

Reply:
@liggitt if you are confused now, it was even more confusing before @MikeSpreitzer made these improvements. There was a fired attribute on the waiter struct which only ever had one value, which was that it hadn't fired. This seems markedly better to me.